chain/blocksync/client.go

```diff
@@ -0,0 +1,445 @@
+		return validRes, nil
+	}
+
+	errString := "doRequest failed for all peers"
+	if singlePeer != nil {
+		errString = fmt.Sprintf("doRequest failed for single peer %s", *singlePeer)
+	}
+	return nil, xerrors.Errorf(errString)
+}
+
+// Process and validate response. Check the status, the integrity of the
+// information returned, and that it matches the request. Extract the information
+// into a `validatedResponse` for the external-facing APIs to select what they
+// need.
+//
+// We are conflating in the single error returned both status and validation
+// errors. Peer penalization should happen here then, before returning, so
+// we can apply the correct penalties depending on the cause of the error.
+func (client *BlockSync) processResponse(
+	req *Request,
+	res *Response,
+	// FIXME: Add the `peer` as argument once we implement penalties.
+) (*validatedResponse, error) {
+	err := res.statusToError()
+	if err != nil {
+		return nil, xerrors.Errorf("status error: %s", err)
+	}
+
+	options := parseOptions(req.Options)
+	if options.noOptionsSet() {
+		// Safety check: this shouldn't have been sent, and even if it did
+		// it should have been caught by the peer in its error status.
+		return nil, xerrors.Errorf("nothing was requested")
+	}
+
+	// Verify that the chain segment returned is in the valid range.
+	// Note that the returned length might be less than requested.
+	resLength := len(res.Chain)
+	if resLength == 0 {
+		return nil, xerrors.Errorf("got no chain in successful response")
+	}
+	if resLength > int(req.Length) {
+		return nil, xerrors.Errorf("got longer response (%d) than requested (%d)",
+			resLength, req.Length)
+	}
+	if resLength < int(req.Length) && res.Status != Partial {
+		return nil, xerrors.Errorf("got less than requested without a proper status: %s", res.Status)
+	}
+
+	validRes := &validatedResponse{}
+	if options.IncludeHeaders {
+		// Check for valid block sets and extract them into `TipSet`s.
+		validRes.tipsets = make([]*types.TipSet, resLength)
+		for i := 0; i < resLength; i++ {
+			validRes.tipsets[i], err = types.NewTipSet(res.Chain[i].Blocks)
+			if err != nil {
+				return nil, xerrors.Errorf("invalid tipset blocks at height (head - %d): %w", i, err)
+			}
+		}
+
+		// Check that the returned head matches the one requested.
+		if !types.CidArrsEqual(validRes.tipsets[0].Cids(), req.Head) {
+			return nil, xerrors.Errorf("returned chain head does not match request")
+		}
+
+		// Check `TipSet`s are connected (valid chain).
+		for i := 0; i < len(validRes.tipsets) - 1; i++ {
+			if validRes.tipsets[i].IsChildOf(validRes.tipsets[i+1]) == false {
+				return nil, fmt.Errorf("tipsets are not connected at height (head - %d)/(head - %d)",
+					i, i+1)
+				// FIXME: Maybe give more information here, like CIDs.
+			}
+		}
+	}
+
+	if options.IncludeMessages {
+		validRes.messages = make([]*CompactedMessages, resLength)
+		for i := 0; i < resLength; i++ {
+			if res.Chain[i].Messages == nil {
```

Comment:
what if there actually arent any messages?

Reply:
you'd have an empty structure but not a nil pointer

Reply:
but that's a good question, i now realize i had the implicit (unfounded) assumption that blocks couldn't be empty